lib/kubernetes-deploy/container_logs.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class ContainerLogs
+    attr_reader :lines, :container_name
+
+    DEFAULT_LINE_LIMIT = 250
+
+    def initialize(parent_id:, container_name:, logger:)
+      @parent_id = parent_id
+      @container_name = container_name
+      @logger = logger
+      @lines = []
+      @last_printed_index = -1
+    end
+
+    def sync(kubectl)
+      new_logs = fetch_latest(kubectl)
+      return unless new_logs.present?
+      @lines += deduplicate(new_logs)
+    end
+
+    def empty?
+      lines.empty?
+    end
+
+    def print_latest(prefix: false)
+      prefix_str = "[#{container_name}]  " if prefix
+      start_at = @last_printed_index + 1
+
+      lines[start_at..-1].each do |msg|
+        @logger.info "#{prefix_str}#{msg}"
+      end
+
+      @last_printed_index = lines.length - 1
+    end
+
+    def print_all
+      lines.each { |line| @logger.info("\t#{line}") }
+    end
+
+    private
+
+    def fetch_latest(kubectl)
+      cmd = ["logs", @parent_id, "--container=#{container_name}", "--timestamps"]
+      cmd << if @last_timestamp.present?
+        "--since-time=#{rfc3339_timestamp(@last_timestamp)}"
+      else
+        "--tail=#{DEFAULT_LINE_LIMIT}"
+      end
+      out, _err, _st = kubectl.run(*cmd, log_failure: false)
```

Comment:
Repost from other PR:

This continues the existing behaviour of treating ALL failures to retrieve logs as "none available I guess". I'm thinking it would just be spammy to output anything about failures because:
(a) log output is for the benefit of the human operator only (in no way affects the control loop);
(b) this call already ends up getting "retried" every polling cycle, which for the runner is every second; and
(c) If Kubernetes is behind on consuming the logs from the container, there's nothing the operator can do about it.
Can you think of a case where it could be helpful to the operator, or a class of error that should actually be handled in some way?

Reply:
> log output is for the benefit of the human operator only

Sadly, this has historically not been true for activefailover at least, where we would sometimes have logic that depends on what is being logged. But I agree with you here in principle and that we should move away from that kind of behaviour.

> thinking it would just be spammy to output anything about failures

Spammy because it might print this over and over in each iteration? I'm imagining a situation where there is not just a "flaky failure" that "fixes itself" but something more permanently wrong. In that case it would be frustrating for the operator to have no idea why nothing is progressing. But I'm ok with this for now.

Is there a timeout for this `kubectl.run` command? Can this theoretically block indefinitely?

Reply:
This comment from your review summary is related to this code:
> For activefailover, this [raising if the pod doesn't log for X seconds] would have saved our ass a few times in the past. Not saying we need this necessarily, but one thing we definitely want to be resilient against is "pod dies and therefore will never log anything, but kubernetes-deploy already started tailing the logs and will never finish". I assume your code here is already better than the code we used to have in activefailover, but having a test specifically for this scenario would be great (not necessarily in this PR, just saying).

With logs outside the control loop, that test case doesn't really make sense. Kubernetes-deploy depends on the exit status of the container to determine its result (although there are some cases where we'll bail even earlier because we've detected configuration errors), so if the pod dies, the task will for sure exit. It will grab and dump any logs that have been available up to that point, which could be all or none of them--it doesn't care.

> Sadly, this has historically not been true for activefailover at least, where we would sometimes have logic that depends on what is being logged. But I agree with you here in principle and that we should move away from that kind of behaviour.

If AF picks up this code though, won't we delete that behaviour? Why would we still need it? And even if we did, I don't see how AF could be using _this copy_ of the log stream.

> Spammy because it might print this over and over in each iteration? I'm imagining a situation where there is not just a "flaky failure" that "fixes itself" but something more permanently wrong.

Yes, and the type of errors I'm anticipating are along the lines of "logs not available yet" or "server not reachable, try again". Can you think of an example of a failure from your experience that would actually be useful to see? (keeping in mind that failures of the pod itself are tracked and reported independently of this)

> Is there a timeout for this kubectl.run command? Can this theoretically block indefinitely?

The timeout is 30 seconds: https://github.com/Shopify/kubernetes-deploy/blob/classes_for_remote_logs/lib/kubernetes-deploy/kubectl.rb#L5. Do you think we should tighten that in general, or maybe for this specific less important command?

That reminds me that kubernetes-run is a heavy user of `Kubeclient` too, and it looks like we haven't configured its timeouts on this repo (I think that means they're 60s). I'll change that too based on your answer.

Reply:
> if the pod dies, the task will for sure exit

👍 

> If AF picks up this code though, won't we delete that behaviour?

yep!

> Can you think of an example of a failure from your experience that would actually be useful to see? (keeping in mind that failures of the pod itself are tracked and reported independently of this)

What we used to do is try to fetch the logs, and on error (any error), just silently try again. I do remember two failure cases specifically: (1) we retry indefinitely because we didn't realize that the reason the logs never appeared is because the container never successfully started in the first place, (2) the kubeclient `watch_pod_logs` method (which uses some kind of HTTP streaming API) never returned and blocked indefinitely (possibly because the container was actually still alive but legitimately stuck and wasn't logging anything? or maybe because of some network fuckery and kubeclient never realized that the connection was already closed?).

(2) doesn't really apply to what we are talking about here. And sounds like you have (1) covered already. So 👍.

> The timeout is 30 seconds

That's the request timeout though, right? So the time after which we give up waiting for the connection to be established? What if the connection is established successfully but then no data is returned for hours (not even a response that says "no logs yet, try again in a bit")? (Somewhat related to my case (2) above).

I'm not sure if this PR (or even this project) is the right level of abstraction to solve this kind of potential problem. So feel free to ignore for now 😄

Reply:
The doc string for that option actually implies it is talking about the whole request, not just connect: `--request-timeout='0': The length of time to wait before giving up on a single server request.` 

I dug in and found the PR that introduced the option, [here](https://github.com/kubernetes/kubernetes/pull/33958/files). As you can see, it is setting `httpClient.Timeout = config.Timeout`, so I looked up that option in the [golang docs](https://golang.org/pkg/net/http/#Client):

```go
        // Timeout specifies a time limit for requests made by this
        // Client. The timeout includes connection time, any
        // redirects, and reading the response body. The timer remains
        // running after Get, Head, Post, or Do return and will
        // interrupt reading of the Response.Body.
```

So I think it really is a timeout on the request completing in its entirety. Do you think 30s is too generous?